url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/decomposition/lu/LUDecompositionNR.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2009-2014, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.dense.decomposition.lu;
+
+import org.ejml.data.DenseMatrix64F;
+
+
+/**
+ * This code is inspired from what's in numerical recipes.
+ *
+ * @author Peter Abeles
+ */
+public class LUDecompositionNR extends LUDecompositionBase_D64 {
+
+    private static final double TINY = 1.0e-40;
+
+
+    /**
+     * <p>
+     * This implementation of LU Decomposition uses the algorithm specified below:
+     *
+     * "Numerical Recipes The Art of Scientific Computing", Third Edition, Pages 48-55<br>
+     * </p>
+     *
+     * @param orig The matrix that is to be decomposed.  Not modified.
+     * @return true If the matrix can be decomposed and false if it can not.  It can
+     * return true and still be singular.
+     */
+    public boolean decompose( DenseMatrix64F orig ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 51-60 into a separate method.